test/powershell/engine/Api/MultithreadedEngine.Tests.ps1

```diff
@@ -0,0 +1,198 @@
+            }
+        }
+
+        It 'cannot invoke a single script asynchronously in the current runspace' {
+            $ps = [powershell]::Create('CurrentRunspace')
+            $sb = {
+                InvokeAsyncHelper -PowerShell $ps -Wait
+            }
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in the current runspace because nested PowerShell instances cannot be
+            # invoked asynchronously
+            $err = { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException' -PassThru
+            GetInnerErrorId -Exception $err.Exception | Should -Be 'InvalidOperation'
+        }
+    }
+
+    Context 'PowerShell::InvokeAsync - Multiple script tests' {
+        It 'can invoke multiple scripts asynchronously' {
+            $sb1 = [scriptblock]::Create("@(1,3,5,7,9,11,13,15,17,19)${sbStub}")
+            $sb2 = [scriptblock]::Create("@(2,4,6,8,10,12,14,16,18,20)${sbStub}")
+            $r1 = [powershell]::Create().AddScript($sb1).InvokeAsync()
+            $r2 = [powershell]::Create().AddScript($sb2).InvokeAsync()
+            [System.Threading.Tasks.Task]::WaitAll(@($r1, $r2))
+            $r1.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r1.IsCompletedSuccessfully | Should -Be $true
+            $r2.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r2.IsCompletedSuccessfully | Should -Be $true
+            $results = @($r1.Result.foreach('Value')) + @($r2.Result.foreach('Value'))
+            Compare-Object -ReferenceObject @(1..20) -DifferenceObject $results -SyncWindow 20 | Should -Be $null
+        }
+    }
+
+    Context 'PowerShell::InvokeAsync - With input and output' {
+        BeforeAll {
+            $d1 = New-Object -TypeName 'System.Management.Automation.PSDataCollection[int]'
+            $d2 = New-Object -TypeName 'System.Management.Automation.PSDataCollection[int]'
+            foreach ($i in 1..20) {
+                $d1.Add($foreach.Current)
+                $foreach.MoveNext() > $null
+                $d2.Add($foreach.Current)
+            }
+            $d1.Complete()
+            $d2.Complete()
+            $script = [scriptblock]::Create("`$input${sbStub}")
+        }
+
+        It 'can invoke multiple scripts asynchronously with input' {
+            $r1 = [powershell]::Create().AddScript($script).InvokeAsync($d1)
+            $r2 = [powershell]::Create().AddScript($script).InvokeAsync($d2)
+            [System.Threading.Tasks.Task]::WaitAll(@($r1, $r2))
+            $r1.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r1.IsCompletedSuccessfully | Should -Be $true
+            $r2.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r2.IsCompletedSuccessfully | Should -Be $true
+            $sortedResults = @($r1.Result) + @($r2.Result) | Sort-Object -Property Time
```

Comment:
I'm a little confused about why sorting using the time property. It seems to be no effect, as `Compare-Object` will do the sorting on those two collections anyway.

Reply:
I had that when I was looking at the results during testing. Removed the unnecessary sorting.